web/middlewares/cors.go

```diff
@@ -9,59 +9,82 @@ import (
 	"github.com/labstack/echo"
 )
 
-// corsBlackList list all routes prefix that are not eligible to CORS
-var corsBlackList = []string{
-	"/auth/",
-}
-
 // MaxAgeCORS is used to cache the CORS header for 12 hours
-var MaxAgeCORS = strconv.Itoa(int(12 * time.Hour / time.Second))
-var allowMethods = strings.Join([]string{echo.GET, echo.HEAD, echo.PUT, echo.PATCH, echo.POST, echo.DELETE}, ",")
+const MaxAgeCORS = "43200"
+
+// CORSOptions contains different options to create a CORS middleware.
+type CORSOptions struct {
+	MaxAge         time.Duration
+	BlackList      []string
+	AllowedMethods []string
+}
 
 // CORS returns a Cross-Origin Resource Sharing (CORS) middleware.
 // See: https://developer.mozilla.org/en/docs/Web/HTTP/Access_control_CORS
-func CORS(next echo.HandlerFunc) echo.HandlerFunc {
-	return func(c echo.Context) error {
+func CORS(opts CORSOptions) echo.MiddlewareFunc {
+	var maxAge string
+	if opts.MaxAge != 0 {
+		maxAge = strconv.Itoa(int(opts.MaxAge.Seconds()))
+	} else {
+		maxAge = MaxAgeCORS
+	}
+
+	var allowedMethods []string
+	if opts.AllowedMethods == nil {
+		allowedMethods = []string{
+			echo.GET,
+			echo.HEAD,
+			echo.PUT,
+			echo.PATCH,
+			echo.POST,
+			echo.DELETE,
+		}
+	}
+
+	allowMethods := strings.Join(allowedMethods, ",")
+
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			req := c.Request()
+			res := c.Response()
 
-		path := c.Path()
-		for _, route := range corsBlackList {
-			if strings.HasPrefix(path, route) {
+			// @TODO validate origin against oauth clients ?
```

Comment:
I think you can remove this TODO. Most of our OAuth clients have no origin (cordova, electron, etc.), but they don't need CORS. CORS is currently used by the web apps (with flat subdomains).